src/modm/platform/spi/stm32/spi_master_dma_impl.hpp.in

```diff
@@ -0,0 +1,185 @@
+modm::platform::SpiMaster{{ id }}_Dma<DmaChannelRx, DmaChannelTx>::transfer(uint8_t data)
+{
+	// this is a manually implemented "fast resumable function"
+	// there is no context or nesting protection, since we don't need it.
+	// there are only two states encoded into 1 bit (LSB of state):
+	//   1. waiting to start, and
+	//   2. waiting to finish.
+	// LSB != Bit0 ?
+	if ( !(state & Bit0) )
+	{
+		// disable DMA for single byte transfer
+		SpiHal{{ id }}::disableInterrupt(SpiBase::Interrupt::TxDmaEnable |
+				SpiBase::Interrupt::RxDmaEnable);
+
+		// wait for previous transfer to finish
+		if (!SpiHal{{ id }}::isTransmitRegisterEmpty())
+			return {modm::rf::Running};
+
+		// start transfer by copying data into register
+		SpiHal{{ id }}::write(data);
+
+		// set LSB = Bit0
+		state |= Bit0;
+	}
+
+	if (!SpiHal{{ id }}::isReceiveRegisterNotEmpty())
+		return {modm::rf::Running};
+
+	SpiHal{{ id }}::read(data);
+
+	// transfer finished
+	state &= ~Bit0;
+	return {modm::rf::Stop, data};
+}
+
+template <class DmaChannelRx, class DmaChannelTx>
+modm::ResumableResult<void>
+modm::platform::SpiMaster{{ id }}_Dma<DmaChannelRx, DmaChannelTx>::transfer(uint8_t *tx,
+		uint8_t *rx, std::size_t length)
+{
+	// this is a manually implemented "fast resumable function"
+	// there is no context or nesting protection, since we don't need it.
+	// there are only two states encoded into 1 bit (Bit1 of state):
+	//   1. initialize index, and
+	//   2. wait for 1-byte transfer to finish.
+
+	// we are only interested in Bit1
+	switch(state & Bit1)
+	{
+	case 0:
+		// we will only visit this state once
+		state |= Bit1;
+		dmaError = false;
+		SpiHal{{ id }}::enableInterrupt(SpiBase::Interrupt::TxDmaEnable |
+				SpiBase::Interrupt::RxDmaEnable);
+
+		if (tx) {
+			Dma::TxChannel::setMemoryAddress(uint32_t(tx));
+			Dma::TxChannel::setMemoryIncrementMode(true);
+		} else {
+			Dma::TxChannel::setMemoryAddress(uint32_t(&dmaDummy));
+			Dma::TxChannel::setMemoryIncrementMode(false);
+		}
+		if (rx) {
+			Dma::RxChannel::setMemoryAddress(uint32_t(rx));
+			Dma::RxChannel::setMemoryIncrementMode(true);
+		} else {
+			Dma::RxChannel::setMemoryAddress(uint32_t(&dmaDummy));
+			Dma::RxChannel::setMemoryIncrementMode(false);
+		}
+
+		Dma::RxChannel::setDataLength(length);
+		dmaReceiveComplete = false;
+		Dma::RxChannel::start();
+
+		Dma::TxChannel::setDataLength(length);
+		dmaTransmitComplete = false;
+		Dma::TxChannel::start();
+
+		[[fallthrough]];
+
+	default:
+		while (true) {
+			if (dmaError)
+				break;
+			if (not dmaTransmitComplete and not dmaReceiveComplete)
+				return { modm::rf::Running };
+			if (SpiHal{{ id }}::getInterruptFlags() & SpiBase::InterruptFlag::FifoTxLevel)
+				return { modm::rf::Running };
+			if (SpiHal{{ id }}::getInterruptFlags() & SpiBase::InterruptFlag::Busy)
+				return { modm::rf::Running };
+			if (SpiHal{{ id }}::getInterruptFlags() & SpiBase::InterruptFlag::FifoRxLevel)
+				return { modm::rf::Running };
```

Comment:
Can't these three be combined?

Reply:
They could be combined, but I rather prefer simple statements. They are easy to understand.